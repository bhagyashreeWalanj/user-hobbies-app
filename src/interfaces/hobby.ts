export enum PassionLevel {
    Low = 'Low',
    Medium = 'Medium',
    High = 'High',
    VeryHigh = 'Very-High',
  }
  
  export interface IHobby {
    id: string
    hobby: string
    passion: PassionLevel | string
    year: string
  }