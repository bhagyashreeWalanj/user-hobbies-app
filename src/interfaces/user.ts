import { IHobby } from './hobby'

// Define the User type
export interface IUser {
    id: string,
      name: string,
      hobbies: IHobby[]
  }