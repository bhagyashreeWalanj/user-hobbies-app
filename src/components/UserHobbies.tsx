import * as React from 'react'
import { useState } from 'react'
import styles from './../styles/UserHobbies.module.scss'
import classnames from 'classnames'

import { IUser } from '../interfaces/user'
import { IHobby, PassionLevel } from '../interfaces/hobby'
import UserHobbiesList from './UserHobbiesList'

export interface UserHobbiesProps {
  user: IUser
  hobbies: IHobby[]
  addHobby(hobby: IHobby): void
  removeHobby(hobbyId: string): void
}
const UserHobbies = ({
  user,
  hobbies,
  addHobby,
  removeHobby,
}: UserHobbiesProps) => {
  const [level, setLevel] = useState(PassionLevel.Low)
  const [hobby, setHobby] = useState('')
  const [year, setYear] = useState('')
  const [error, setError] = useState('')

  const handleNewHobby = () => {
    if (
      hobbies.find((h) => h.hobby.toLowerCase() === hobby.trim().toLowerCase())
    ) {
      setError(`The hobby "${hobby}" already exists in the hobbies list`)
      return
    } else if (hobby.trim() === '') {
      setError('Please enter your hobby')
      return
    } else if (year.trim() === '') {
      setError('Please enter year')
      return
    } else if (!/^[0-9]+$/.test(year)) {
      setError('Please enter correct year')
      return
    } else {
      addHobby({
        id: (Math.floor(Math.random() * 100) + 8).toString(),
        hobby: hobby,
        passion: level,
        year: year.toString(),
      })
      setHobby('')
      setYear('')
      setError('')
    }
  }

  if (!user) {
    return (
      <div className={classnames(styles.container, styles.containerEmpty)}>
        <div className={styles.emptyContainer}>
          <span className={styles.noUserDiv}>No Data Selected !!</span>
        </div>
      </div>
    )
  }

  return (
    <div className={styles.container}>
      <div className={styles.errorDiv}>{error}</div>
      <div className={styles.inputContainer}>
        <label className={styles.inputLabel}>Passion-Level : </label>
        <select
          className={styles.inputContainerField}
          value={level}
          onChange={({ target: { value } }) => {
            setLevel(value as PassionLevel)
            setError('')
          }}
        >
          {Object.values(PassionLevel).map((lvl) => (
            <option key={lvl} value={lvl}>
              {lvl}
            </option>
          ))}
        </select>
        <label className={styles.inputLabel}>Hobby:</label>
        <input
          type="text"
          value={hobby}
          id="hobby"
          onChange={({
            target: { value },
          }: React.ChangeEvent<HTMLInputElement>) => setHobby(value)}
          placeholder="Enter user hobby"
          className={styles.inputContainerField}
        />
        <label className={styles.inputLabel}>Year : </label>
        <input
          type="text"
          maxLength={4}
          value={year}
          id="year"
          onChange={({
            target: { value },
          }: React.ChangeEvent<HTMLInputElement>) => setYear(value)}
          placeholder="Enter year"
          className={styles.inputContainerField}
        />
        <button className={styles.addHobbybutton} onClick={handleNewHobby}>
          Add Hobby
        </button>
      </div>
      <div className={styles.itemHeader}>
        <span>Passion-Level</span>
        <span>Hobby</span>
        <span>Year</span>
      </div>
      <UserHobbiesList hobbies={hobbies} removeHobby={removeHobby} />
    </div>
  )
}

export default UserHobbies
