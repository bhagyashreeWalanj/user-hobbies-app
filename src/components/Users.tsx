import * as React from 'react'
import { useState, useEffect } from 'react'
import styles from './../styles/Users.module.scss'
import { IUser } from './../interfaces/user'
import UserList from './UserList'

export interface IProps {
  Users: IUser[]
  getAllUsers(): void
  selectedUserId: string
  selectUser(id: string): void
  addUser(id: string): void
}

const Users = (props: IProps) => {
  const [newUserName, setNewUserName] = useState('')
  const { Users, selectedUserId, selectUser, addUser, getAllUsers } = props
  const [error, setError] = useState('')

  const onAddUser = () => {
    if (newUserName !== '') {
      addUser(newUserName)
      setNewUserName('')
    } else {
      setError('Enter user name')
      return
    }
  }

  useEffect(() => {
    getAllUsers()
  }, [getAllUsers])
  return (
    <div>
      <div className={styles.errorDiv}>{error}</div>
      <div className={styles.addUserContainer}>
        <input
          type="text"
          id="newUser"
          placeholder="Enter user name"
          value={newUserName}
          onChange={({
            target: { value },
          }: React.ChangeEvent<HTMLInputElement>) => setNewUserName(value)}
          className={styles.searchContainerField}
        />
        <button
          className={styles.addUserbutton}
          onClick={onAddUser}
          id="addNewbutton"
        >
          Add User
        </button>
      </div>

      <UserList
        Users={Users}
        selectedUserId={selectedUserId}
        selectUser={selectUser}
      />
    </div>
  )
}

export default Users
