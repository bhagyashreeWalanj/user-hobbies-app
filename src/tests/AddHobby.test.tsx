import React from 'react';
import { mount, shallow, render } from 'enzyme';
import Users, { IProps } from '../components/Users';
import UserHobbies from '../components/UserHobbies';
import { IUser } from '../interfaces/user';


 describe('Testing Add Hobby Component', () => {
    const addHobby = jest.fn();
    const removeHobby = jest.fn();
  const users:IUser ={
    id: '1',
    name: 'John Hecker',
    hobbies: [{
      id: '2',
      hobby: 'Playing Cricket',
      passion: 'Low',
      year: '1990'
    }]
  };
  

  const wrapper = shallow(<UserHobbies user={users} hobbies={[]} addHobby={addHobby} removeHobby={removeHobby} />);
    const hobbyInput = wrapper.find('#hobby')
    const yearInput = wrapper.find('#year');


  it('Renders user Hobbies Component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should display error on empty Hobby field', () => {
    hobbyInput.simulate('change', { target: { value: '' } });
    wrapper.find('.addHobbybutton').simulate('click');
    const errorDiv = wrapper.find('.errorDiv')
    const errorText = errorDiv.text();
    expect(errorText).toEqual('Please enter your hobby');
  });

  it('should display error on empty Year input field', () => {
    hobbyInput.simulate('change', { target: { value: 'Sam' } });
    yearInput.simulate('change', { target: { value: '' } });
    wrapper.find('.addHobbybutton').simulate('click');
    const errorDiv = wrapper.find('.errorDiv')
    const errorText = errorDiv.text();
    expect(errorText).toEqual('Please enter year');
  });

  it('should get added in hobbies list', () => {
    hobbyInput.simulate('change', { target: { value: 'Play Basketball' } });
    yearInput.simulate('change', { target: { value: '1990' } });
    wrapper.find('.addHobbybutton').simulate('click');
    expect(addHobby).toBeCalledTimes(1);
  });
 });




