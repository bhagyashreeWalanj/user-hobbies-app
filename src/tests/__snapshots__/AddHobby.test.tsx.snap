// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Add Hobby Component Renders user Hobbies Component 1`] = `
<div
  className="container"
>
  <div
    className="errorDiv"
  />
  <div
    className="inputContainer"
  >
    <label
      className="inputLabel"
    >
      Passion-Level : 
    </label>
    <select
      className="inputContainerField"
      onChange={[Function]}
      value="Low"
    >
      <option
        key="Low"
        value="Low"
      >
        Low
      </option>
      <option
        key="Medium"
        value="Medium"
      >
        Medium
      </option>
      <option
        key="High"
        value="High"
      >
        High
      </option>
      <option
        key="Very-High"
        value="Very-High"
      >
        Very-High
      </option>
    </select>
    <label
      className="inputLabel"
    >
      Hobby:
    </label>
    <input
      className="inputContainerField"
      id="hobby"
      onChange={[Function]}
      placeholder="Enter user hobby"
      type="text"
      value=""
    />
    <label
      className="inputLabel"
    >
      Year : 
    </label>
    <input
      className="inputContainerField"
      id="year"
      maxLength={4}
      onChange={[Function]}
      placeholder="Enter year"
      type="text"
      value=""
    />
    <button
      className="addHobbybutton"
      onClick={[Function]}
    >
      Add Hobby
    </button>
  </div>
  <div
    className="itemHeader"
  >
    <span>
      Passion-Level
    </span>
    <span>
      Hobby
    </span>
    <span>
      Year
    </span>
  </div>
  <UserHobbiesList
    hobbies={Array []}
    removeHobby={[MockFunction]}
  />
</div>
`;
